### PR TITLE
Update FormatType in C++

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR removes the DefaultFormat enumerator in C++. It also removes _instance from OutputStream in C++.

It's a partial C++ equivalent of #2594.

### CODE REVIEW

File: cpp/src/Ice/TraceUtil.h [hunk 1]
```diff
@@ -16,9 +17,10 @@ namespace Ice
 
 namespace IceInternal
 {
-    void traceSend(const Ice::OutputStream&, const Ice::LoggerPtr&, const TraceLevelsPtr&);
+    class Instance;
+
+    void traceSend(const Ice::OutputStream&, const InstancePtr& instance, const Ice::LoggerPtr&, const TraceLevelsPtr&);
     void traceRecv(const Ice::InputStream&, const Ice::LoggerPtr&, const TraceLevelsPtr&);
-    void trace(const char*, const Ice::OutputStream&, const Ice::LoggerPtr&, const TraceLevelsPtr&);
```
Comment (bernardnormier) [inline cpp/src/Ice/TraceUtil.h:21]: Was not used.

File: php/lib/Ice/Value.php [hunk 2]
```diff
@@ -66,11 +66,4 @@ class SliceInfo
     public $isLastSlice;
 }
 
-class FormatType
```
Comment (bernardnormier) [inline php/lib/Ice/Value.php:69]: Removed because unused.

File: cpp/src/slice2cpp/CPlusPlusUtil.cpp [hunk 3]
```diff
@@ -607,7 +607,7 @@ Slice::opFormatTypeToString(const OperationPtr& op)
     switch (op->format())
     {
         case DefaultFormat:
-            return "::Ice::FormatType::DefaultFormat";
+            return "::std::nullopt";
```
Comment (bernardnormier) [inline cpp/src/slice2cpp/CPlusPlusUtil.cpp:610]: In the slice2xxx compilers, FormatType has still 3 values that correspond to the metadata.

File: cpp/src/Ice/OutputStream.cpp [hunk 4]
```diff
@@ -162,11 +161,11 @@ void
 Ice::OutputStream::initialize(Instance* instance, const EncodingVersion& encoding)
 {
     assert(instance);
-
-    _instance = instance;
     _encoding = encoding;
 
-    _format = _instance->defaultsAndOverrides()->defaultFormat;
+    _format = instance->defaultsAndOverrides()->defaultFormat;
```
Comment (pepone) [inline cpp/src/Ice/OutputStream.cpp:166]: Shouldn't this be a second constructor? it is not clear why we use a `initialize` method.

**Review by pepone — APPROVED** [review]

**Review by externl — APPROVED** [review]
Looks good!

File: cpp/src/Ice/OutputStream.cpp (hunk 4 above)
Comment (bernardnormier) [reply cpp/src/Ice/OutputStream.cpp:166]: I will clean this up in a follow-up PR.

`initialize` is used by the IceDB template code, but I think we can remove it.

**Pull request merged**